src/bp/nlu-core/training-worker-queue.ts

```diff
@@ -0,0 +1,253 @@
+    const worker = cluster.workers[msg.destWid!]
+    worker?.send(msg) // TODO: find out why this is sometimes undefined.
+  }
+
+  function killTrainingWorker(msg: OutgoingMessage) {
+    const worker = cluster.workers[msg.destWid!]
+
+    if (!worker) {
+      const response: IncomingMessage = { type: 'training_canceled', payload: {}, srcWid: msg.destWid! }
+      return sendToWebWorker(response)
+    }
+
+    worker!.kill('SIGKILL')
+    const exitHandler = (worker: Worker, _exitCode: number, _signal: string) => {
+      const response: IncomingMessage = { type: 'training_canceled', payload: {}, srcWid: worker.id }
+      sendToWebWorker(response)
+    }
+    cluster.once('exit', exitHandler)
+  }
+
+  registerMsgHandler('make_new_worker', async (msg: OutgoingMessage) => spawnNewTrainingWorker(msg.payload.config!))
+  registerMsgHandler('cancel_training', killTrainingWorker)
+  registerMsgHandler('start_training', sendToTrainingWorker)
+
+  registerMsgHandler('log', sendToWebWorker)
+  registerMsgHandler('worker_ready', sendToWebWorker)
+  registerMsgHandler('training_done', sendToWebWorker)
+  registerMsgHandler('training_progress', sendToWebWorker)
+  registerMsgHandler('training_error', sendToWebWorker)
+}
+
+if (cluster.isWorker && process.env.WORKER_TYPE === WORKER_TYPES.TRAINING) {
+  const config = JSON.parse(process.env.NLU_CONFIG!)
+
+  const srcWid = cluster.worker.id
+  const logger: NLU.Logger = {
+    info: (info: string) => {
+      const msg: IncomingMessage = { type: 'log', payload: { log: { info } }, srcWid }
+      process.send!(msg)
+    },
+    warning: (warning: string) => {
+      const msg: IncomingMessage = { type: 'log', payload: { log: { warning } }, srcWid }
+      process.send!(msg)
+    },
+    error: (error: string) => {
+      const msg: IncomingMessage = { type: 'log', payload: { log: { error } }, srcWid }
+      process.send!(msg)
+    }
+  }
+
+  async function main() {
+    const tools = await initializeTools(config, logger)
+    process.on('message', async (msg: OutgoingMessage) => {
+      if (msg.type === 'start_training') {
+        const { input } = msg.payload
+
+        const progressCb = (progress: number) => {
+          const res: IncomingMessage = { type: 'training_progress', payload: { progress }, srcWid }
+          process.send!(res)
+        }
+
+        let output: TrainOutput | undefined
+        try {
+          output = await Trainer(input!, tools, progressCb)
+        } catch (err) {
+          const res: IncomingMessage = { type: 'training_error', payload: { error: err.message }, srcWid }
+          process.send!(res)
+        }
+
+        const res: IncomingMessage = { type: 'training_done', payload: { output }, srcWid }
+        process.send!(res)
+      }
+    })
+
+    const res: IncomingMessage = { type: 'worker_ready', payload: {}, srcWid }
+    process.send!(res)
+  }
+
+  // tslint:disable-next-line: no-floating-promises
+  main()
```

Comment:
why wrapping all this in a function to add a tslint disabled. 

Here's my suggestion but you don't have to take it :

```ts
async function onMessage(tools: Tools, msg: OutgoingMessage){
      if (msg.type === 'start_training') {
        const { input } = msg.payload

        const progressCb = (progress: number) => {
          const res: IncomingMessage = { type: 'training_progress', payload: { progress }, srcWid }
          process.send!(res)
        }

        let output: TrainOutput | undefined
        try {
          output = await Trainer(input!, tools, progressCb)
        } catch (err) {
          const res: IncomingMessage = { type: 'training_error', payload: { error: err.message }, srcWid }
          process.send!(res)
        }

        const res: IncomingMessage = { type: 'training_done', payload: { output }, srcWid }
        process.send!(res)
      }
    })

    const res: IncomingMessage = { type: 'worker_ready', payload: {}, srcWid }
    process.send!(res)
  }}
}

initializeTools(config, logger)
  .then(tools => process.on('message', onMessage.bind(null, tools))
  .catch(err => /* handle error properly here*/)
```

Reply:
:+1: